### ISSUE TITLE
Bug: The elements of the batch contain different keys. Cannot batch them ...

### ISSUE THREAD

**Issue by asusdisciple** [huggingface/transformers#33404]
### System Info

DGX Station
Ubuntu 20.04
Transformers 4.44.2
torch 2.4.1

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

So I ran into this bug which I did not experience in prior versions of transformers (Unfortunately I did not document which was the last one working). What I basically do in transcribe_audio() is to call different model architectures depending on their name:

```
def transcribe_audio(audiopath: List[str], model, repo_name: str, parameters: {}, lang: str) -> List[str]:
  tmp = model(audiopath,
                      generate_kwargs={"language": lang}
                      )
  
          res = [i["text"] for i in tmp]
          return res

```
The bug:

```
You have passed task=transcribe, but also have set `forced_decoder_ids` to [[1, None], [2, 50359]] which creates a conflict. `forced_decoder_ids` will be ignored in favor of task=transcribe.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.

Traceback (most recent call last):
  File "/raid/.../Whisper/main.py", line 139, in <module>
    transcript = transcribe_audio(filenames, m, repo, par, lang.value)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/benchmark.py", line 247, in transcribe_audio
    tmp = model(audiopath,
          ^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 284, in __call__
    return super().__call__(inputs, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/transformers/pipelines/base.py", line 1238, in __call__
    outputs = list(final_iterator)
              ^^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/transformers/pipelines/pt_utils.py", line 124, in __next__
    item = next(self.iterator)
           ^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/transformers/pipelines/pt_utils.py", line 269, in __next__
    processed = self.infer(next(self.iterator), **self.params)
                           ^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/torch/utils/data/dataloader.py", line 630, in __next__
    data = self._next_data()
           ^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/torch/utils/data/dataloader.py", line 673, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/torch/utils/data/_utils/fetch.py", line 43, in fetch
    return self.collate_fn(data)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/.../Whisper/venv2/lib/python3.12/site-packages/transformers/pipelines/base.py", line 175, in inner
    raise ValueError(
ValueError: The elements of the batch contain different keys. Cannot batch them ({'input_features', 'is_last'} != {'num_frames', 'input_features', 'is_last'})
```

This error is only raised when I initialize a new model. So I loop through different models, whisper-v2, whisper-v3, ... and initialize a new model at the beginng of each loop. First loop with inference runs through, then in the second loop after a few inferences this bug appears. Do you have an idea what this can relate to? I just overwrite my old model object with the new one. I also empty any torch cache before the new init.

### Expected behavior

Just to work like before, looping through models and doing inference. There should not be any internal bug like this.

**Comment by minmie** [comment]
I have the same problem. 
when i use pipe to inference with batch_size=1, everything is ok. However the error occur when infer with batch_size>1.

model: whisper-large-v3-zh-punct
autio_data: wav data

```python

import time

from transformers import pipeline, WhisperForConditionalGeneration, AutoModelForSpeechSeq2Seq, AutoProcessor
import os
import torch

DATA_DIR = r'C:\Users\chenjq2\Desktop\wav格式录音'
# DATA_DIR = r'./test_data'
LANGUAGE = 'zh'
TASK = 'transcribe'
files = os.listdir(DATA_DIR)

paths = []
for name in files:
    paths.append(os.path.join(DATA_DIR, name))
MODEL_ID = r"C:\Users\chenjq2\Desktop\tmp\models--BELLE-2--Belle-whisper-large-v3-zh-punct\snapshots\f81f1ac2f123f118094a7baa69e532eab375600e"
device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32
model = AutoModelForSpeechSeq2Seq.from_pretrained(
    MODEL_ID, torch_dtype=torch_dtype, low_cpu_mem_usage=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(MODEL_ID, language=LANGUAGE, task=TASK)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)
t1 = time.time()
print(pipe(paths, batch_size=4))
print(f'time cost:{time.time()-t1}')

```
error msg:
```
E:\program\anaconda3\envs\nlp\lib\site-packages\torch\_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()
Traceback (most recent call last):
  File "E:\program\anaconda3\envs\nlp\lib\site-packages\torch\utils\data\dataloader.py", line 630, in __next__
    data = self._next_data()
  File "E:\program\anaconda3\envs\nlp\lib\site-packages\torch\utils\data\dataloader.py", line 674, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "E:\program\anaconda3\envs\nlp\lib\site-packages\torch\utils\data\_utils\fetch.py", line 42, in fetch
    return self.collate_fn(data)
  File "E:\program\anaconda3\envs\nlp\lib\site-packages\transformers\pipelines\base.py", line 175, in inner
    raise ValueError(
ValueError: The elements of the batch contain different keys. Cannot batch them ({'num_frames', 'input_features', 'is_last'} != {'input_features', 'is_last'})
```




The differnece is due to this:

It will have an additional field num_frames if the code runs to block 2, but not if it runs to block 1.


XXX\transformers\pipelines\automatic_speech_recognition.py
![Snipaste_2024-09-11_10-40-47](https://github.com/user-attachments/assets/6b50c8d9-97e3-4e7e-8aa5-dccf3a48fe3b)

**Comment by asusdisciple** [comment]
Hello I found the root of the problem:

I used a` transformers pipeline `with a `WhisperForConditionalGeneration` model and explicitly set the `batch_size` parameter.
However when you set the` batch_size` parameter without setting` chunk_length_s `this error occurs. After setting `chunk_length_s` to some value the error did not appear anymore.

I thinks this should not be the case right? From what I understood not using the chunk_length_s parameter should result in using default sequential batching: https://huggingface.co/blog/asr-chunking

Or is this only true for the ASR pipeline? Should I use that instead of the vanilla pipeline or isnt ASR inferred by the model type anyway?

**Comment by ylacombe** [comment]
Hey @asusdisciple, thanks for opening this issue and digging in to find the root of the problem!
Do you think you could write a **small* self-sufficient and reproducible code, so that I can better identify and correct the root of the issue ?
Thanks!

**Comment by ylacombe** [comment]
For reference, can't seem to be able to reproduce, more details [here](https://github.com/huggingface/transformers/issues/33415#issuecomment-2345587578)

**Comment by ylacombe** [comment]
Hey @asusdisciple, I was able to reproduce this issue! More details [here](https://github.com/huggingface/transformers/issues/33415#issuecomment-2353433444)!

TL;DR: the bug happens when, in the same batch, there are samples that are longer and shorter than 30s, which is the fixed length on which Whisper was trained. You can already avoid this bug by passing chunk_length_s=30.0 to the pipeline, as you rightfully indicated.

**Comment by asusdisciple** [comment]
Thanks for sticking to the topic @ylacombe ! Do you think this should be fixed or addressed with a more detailed warning message?
